### PR TITLE
Add testing labs

### LINKED ISSUE #319: add model and crud operations for testing labs
Add model and CRUD API operations for testing labs. 

Fields
 - name, address, pincode, phone number
 - local body, district and state
 - capacity of tests per day (0 default => no info)
 
Permissions
  - read -> all verified users
  - write -> any one who has permission >= DistrictLabAdmin

### PR DESCRIPTION
**Pull request by abhiabhi94**
Add testing labs model and API for create, read, update and delete options.

TODO:
- Add tests for this API

Fixes #319

### CODE REVIEW

**Comment by abhiabhi94** [comment]
If someone could review #298, I can confidently write tests for this API without worrying about merge conflicts.

File: care/facility/api/viewsets/testing_lab.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from django_filters import rest_framework as filters
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.viewsets import ModelViewSet
+
+from care.facility.api.serializers.testing_lab import TestingLabSerializer
+from care.facility.models.testing_lab import TestingLab
+
+
+class TestingLabFilterSet(filters.FilterSet):
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:9]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    testing_capacity_per_day = models.IntegerField(default=0, blank=True)
+    created_by = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
+
+    @staticmethod
+    def has_object_read_permissions(self, request):
+        """Allow all verified users to view"""
+        return self.request.user.verified
+
+    @staticmethod
+    def has_object_write_permissions(self, request):
```
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:32]: Codacy Issue found: [Unused argument 'self'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+
+from care.facility.models.testing_lab import TestingLab
+
+
+class TestingLabSerializer(serializers.ModelSerializer):
+    class Meta:
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:7]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+from django_filters import rest_framework as filters
+from rest_framework.permissions import IsAuthenticated
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:2]: Codacy Issue found: [Unable to import 'rest_framework.permissions'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py (hunk 2 above)
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:32]: Codacy Issue found: [Static method with 'self' as first argument](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from django_filters import rest_framework as filters
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:1]: Codacy Issue found: [Unable to import 'django_filters'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    operations = [
+        migrations.CreateModel(
+            name='TestingLab',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('name', models.CharField(max_length=200)),
+                ('address', models.TextField()),
+                ('pincode', models.IntegerField(validators=[django.core.validators.RegexValidator(code='invalid_pincode', message='Please enter a valid pincode.', regex='^((\\d{6}$')])),
+                ('phone_number', models.CharField(blank=True, max_length=14, validators=[django.core.validators.RegexValidator(code='invalid_mobile', message='Please Enter 10/11 digit mobile number or landline as 0<std code><phone number>', regex='^((\\+91|91|0)[\\- ]{0,1})?[456789]\\d{9}$')])),
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab.py:23]: Codacy Issue found: [Line too long (296/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='TestingLab',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('name', models.CharField(max_length=200)),
+                ('address', models.TextField()),
+                ('pincode', models.IntegerField(validators=[django.core.validators.RegexValidator(code='invalid_pincode', message='Please enter a valid pincode.', regex='^((\\d{6}$')])),
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab.py:22]: Codacy Issue found: [Line too long (186/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.viewsets import ModelViewSet
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:3]: Codacy Issue found: [Unable to import 'rest_framework.viewsets'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0091_auto_20200413_1714.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('facility', '0090_testinglab_created_by'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='testinglab',
+            name='pincode',
+            field=models.IntegerField(validators=[django.core.validators.RegexValidator(code='invalid_pincode', message='Please enter a valid pincode.', regex='^(\\d{6}$)')]),
```
Comment (vigneshhari) [inline care/facility/migrations/0091_auto_20200413_1714.py:17]: Codacy Issue found: [Line too long (175/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0090_testinglab_created_by.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+        ('facility', '0089_testinglab'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='testinglab',
+            name='created_by',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to=settings.AUTH_USER_MODEL),
```
Comment (vigneshhari) [inline care/facility/migrations/0090_testinglab_created_by.py:19]: Codacy Issue found: [Line too long (134/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    local_body = filters.CharFilter(field_name="local_body_id")
+    district = filters.CharFilter(field_name="district_id")
+    state = filters.CharFilter(field_name="state_id")
+
+    local_body_name = filters.CharFilter(field_name="local_body__name", lookup_expr="icontains")
+    district_name = filters.CharFilter(field_name="district__name", lookup_expr="icontains")
+    state_name = filters.CharFilter(field_name="state__name", lookup_expr="icontains")
+
+
+class TestingLabViewSet(ModelViewSet):
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:22]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
+
+from care.facility.models.testing_lab import TestingLab
+
+
+class TestingLabSerializer(serializers.ModelSerializer):
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:6]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                ('phone_number', models.CharField(blank=True, max_length=14, validators=[django.core.validators.RegexValidator(code='invalid_mobile', message='Please Enter 10/11 digit mobile number or landline as 0<std code><phone number>', regex='^((\\+91|91|0)[\\- ]{0,1})?[456789]\\d{9}$')])),
+                ('testing_capacity_per_day', models.IntegerField(blank=True, default=0)),
+                ('district', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='users.District')),
+                ('local_body', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='users.LocalBody')),
+                ('state', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='users.State')),
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab.py:27]: Codacy Issue found: [Line too long (132/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 14]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:1]: Codacy Issue found: [Unable to import 'rest_framework'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    pincode = models.IntegerField(validators=[pincode_regex])
+    phone_number = models.CharField(max_length=14, blank=True, validators=[phone_number_regex])
+    local_body = models.ForeignKey(LocalBody, on_delete=models.SET_NULL, null=True, blank=True)
+    district = models.ForeignKey(District, on_delete=models.SET_NULL, null=True, blank=True)
+    state = models.ForeignKey(State, on_delete=models.SET_NULL, null=True, blank=True)
+    testing_capacity_per_day = models.IntegerField(default=0, blank=True)
+    created_by = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
+
+    @staticmethod
+    def has_object_read_permissions(self, request):
```
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:27]: Codacy Issue found: [Static method with 'self' as first argument](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        migrations.CreateModel(
+            name='TestingLab',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('name', models.CharField(max_length=200)),
+                ('address', models.TextField()),
+                ('pincode', models.IntegerField(validators=[django.core.validators.RegexValidator(code='invalid_pincode', message='Please enter a valid pincode.', regex='^((\\d{6}$')])),
+                ('phone_number', models.CharField(blank=True, max_length=14, validators=[django.core.validators.RegexValidator(code='invalid_mobile', message='Please Enter 10/11 digit mobile number or landline as 0<std code><phone number>', regex='^((\\+91|91|0)[\\- ]{0,1})?[456789]\\d{9}$')])),
+                ('testing_capacity_per_day', models.IntegerField(blank=True, default=0)),
+                ('district', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='users.District')),
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab.py:25]: Codacy Issue found: [Line too long (138/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+                ('local_body', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='users.LocalBody')),
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab.py:26]: Codacy Issue found: [Line too long (141/120)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+from django.utils.translation import gettext_lazy as _
+
+from care.facility.models import phone_number_regex
+from care.users.models import District, LocalBody, State
+
+User = get_user_model()
+
+
+class TestingLab(models.Model):
+    pincode_regex = RegexValidator(
```
[the remark this answers is not in the document]
Comment (abhiabhi94) [reply care/facility/models/testing_lab.py:13]: I had initially put it there, but pre-commit flake started behaving weirdly. 
The earlier validators put inside the file were causing an issue. So just to hack around the problem I transferred the code here.

File: care/facility/models/testing_lab.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (abhiabhi94) [reply care/facility/models/testing_lab.py:27]: Do you mean this class https://github.com/coronasafe/care/blob/master/care/facility/models/mixins/permissions/base.py#L4
It has a lot of different permission rules which are different from the ones told earlier for the `testing_lab` model.

File: care/facility/models/testing_lab.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    pincode = models.IntegerField(validators=[pincode_regex])
+    phone_number = PhoneNumberField()
+    local_body = models.ForeignKey(LocalBody, on_delete=models.SET_NULL, null=True, blank=True)
+    district = models.ForeignKey(District, on_delete=models.SET_NULL, null=True, blank=True)
+    state = models.ForeignKey(State, on_delete=models.SET_NULL, null=True, blank=True)
+    testing_capacity_per_day = models.IntegerField(default=0, blank=True)
+    created_by = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
+
+    @staticmethod
+    def has_object_read_permissions(self, request):
```
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:27]: Codacy Issue found: [Static method with 'self' as first argument](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py [hunk 20]
```diff
@@ -0,0 +1,34 @@
+from django.contrib.auth import get_user_model
+from django.core.validators import RegexValidator
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from phonenumber_field.modelfields import PhoneNumberField
```
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:5]: Codacy Issue found: [Unable to import 'phonenumber_field.modelfields'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/migrations/0089_testinglab_squashed_0092_auto_20200414_2205.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
+# Generated by Django 2.2.11 on 2020-04-14 16:36
+
+from django.conf import settings
+import django.core.validators
+from django.db import migrations, models
+import django.db.models.deletion
+import phonenumber_field.modelfields
```
Comment (vigneshhari) [inline care/facility/migrations/0089_testinglab_squashed_0092_auto_20200414_2205.py:7]: Codacy Issue found: [Unable to import 'phonenumber_field.modelfields'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+
+from care.facility.models.testing_lab import TestingLab
+from care.utils.serializer.phonenumber_ispossible_field import PhoneNumberIsPossibleField
+
+
+class TestingLabSerializer(serializers.ModelSerializer):
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:7]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py (hunk 5 above)
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:1]: Codacy Issue found: [Unable to import 'django_filters'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py (hunk 8 above)
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:3]: Codacy Issue found: [Unable to import 'rest_framework.viewsets'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+    phone_number = PhoneNumberIsPossibleField()
+
+    class Meta:
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:10]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/serializers/testing_lab.py [hunk 24]
```diff
@@ -0,0 +1,12 @@
+from rest_framework import serializers
```
Comment (vigneshhari) [inline care/facility/api/serializers/testing_lab.py:1]: Codacy Issue found: [Unable to import 'rest_framework'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/models/testing_lab.py [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    testing_capacity_per_day = models.IntegerField(default=0, blank=True)
+    created_by = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
+
+    @staticmethod
+    def has_object_read_permissions(self, request):
+        """Allow all verified users to view"""
+        return self.request.user.verified
+
+    @staticmethod
+    def has_object_write_permissions(self, request):
```
Comment (vigneshhari) [inline care/facility/models/testing_lab.py:32]: Codacy Issue found: [Static method with 'self' as first argument](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

Comment (vigneshhari) [inline care/facility/models/testing_lab.py:32]: Codacy Issue found: [Unused argument 'self'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py (hunk 1 above)
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:9]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py (hunk 4 above)
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:2]: Codacy Issue found: [Unable to import 'rest_framework.permissions'](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

File: care/facility/api/viewsets/testing_lab.py [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+    local_body = filters.NumberFilter(field_name="local_body_id")
+    district = filters.NumberFilter(field_name="district_id")
+    state = filters.NumberFilter(field_name="state_id")
+
+    local_body_name = filters.CharFilter(field_name="local_body__name", lookup_expr="icontains")
+    district_name = filters.CharFilter(field_name="district__name", lookup_expr="icontains")
+    state_name = filters.CharFilter(field_name="state__name", lookup_expr="icontains")
+
+
+class TestingLabViewSet(ModelViewSet):
```
Comment (vigneshhari) [inline care/facility/api/viewsets/testing_lab.py:22]: Codacy Issue found: [Too few public methods (0/2)](https://app.codacy.com/gh/coronasafe/care/pullRequest?prid=5348729)

**Comment by abhiabhi94** [comment]
This is no longer required, #334 handles the issue as of now.

**Pull request closed without merging**